### PR TITLE
server: update cache_prompt documentation [no ci]

### LINKED ISSUE #7594: Bug: [Server] Prompt caching causes subsequent identical requests to return different token probabilities
### What happened?

Hi there,

I noticed that prompt caching on the server causes subsequent identical requests to return different token probabilities, even though I use just one slot. When using greedy decoding (`temperature=-1.0`), I would expect the results to be exactly identical when using one server slot.

### Server parameters
```
./server -m ../local_models/orca-2-7b.Q4_K_M.gguf -c 4096 -ngl 1000 -np 1 -cb -v
```

### Request/Response 1
```
// Request 1
curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '{"prompt": "Test", "id_slot": 0, "n_predict": 1, "n_probs": 1, "samplers": ["temperature"], "seed": 1234, "temperature": -1.0, "cache_prompt": true}'
```

```
// Response 1
{
    "content": "imon",
    "tokens_predicted": 1,
    "tokens_evaluated": 2,
    "prompt": "Test",
    "tokens_cached": 2,
    "timings":
    {
        "prompt_n": 2,
    },
    "completion_probabilities":
    [
        {
            "content": "imon",
            "probs":
            [
                {
                    "tok_str": "imon",
                    "prob": 0.5299757719039917
                }
            ]
        }
    ]
}
```

### Request/Response 2
```
// Request 2, identical to request 1
curl --request POST \
    --url http://localhost:8080/completion \
    --header "Content-Type: application/json" \
    --data '{"prompt": "Test", "id_slot": 0, "n_predict": 1, "n_probs": 1, "samplers": ["temperature"], "seed": 1234, "temperature": -1.0, "cache_prompt": true}'
```

```
// Response 2
{
    "content": "imon",
    "tokens_predicted": 1,
    "tokens_evaluated": 2,
    "prompt": "Test",
    "tokens_cached": 2,
    "timings":
    {
        "prompt_n": 1,
    },
    "completion_probabilities":
    [
        {
            "content": "imon",
            "probs":
            [
                {
                    "tok_str": "imon",
                    "prob": 0.5112787485122681
                }
            ]
        }
    ]
}
```

I don't know how this is implemented internally. Does the caching affect the batch size, which in turn affects rounding?

Best
Leon

### Name and Version

version: 3024 (2b737caa)
built with Apple clang version 15.0.0 (clang-1500.3.9.4) for arm64-apple-darwin23.5.0

### What operating system are you seeing the problem on?

Mac

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by JohannesGaessler**
Fixes https://github.com/ggerganov/llama.cpp/issues/7594 .

Generally speaking it is not possible to guarantee bit-for-bit identical results when the batch size is varied. As a consequence `cache_prompt` that caches not only the actual prompt (batch size >> 1) but also the generated tokens (batch size 1) cannot guarantee bit-for-bit identical results. This PR simply updates the documentation to inform users of this.

### CODE REVIEW

**Comment by steampunque** [comment]
I think there is a very subtle thing going on here which needs to be clarified.  With no prompt cache the prompt goes through in chunks of batch size with the final chunk potentially less than the batch size where the first token can simultaneously be generated with the final prompt process.  When resubmitting the identical prompt the same results will always be achieved.  When prompt cache is turned on, the first generation will be identical to that without prompt cache, but a following run with the identical prompt will start generating immediately with a batch size of 1 (it will not be 1 + residual prompt process for the first generation).  So it seems like it should be possible to not start generation on the final partial batch with prompt process so token generation is always done with batch size 1 whether prompt cache is on or off to get the results identical with or without prompt cache.

**Comment by steampunque** [comment]
Here is an example illustrating the point I made in my last comment.  Using mistral instruct v0.3 with the simple prompt "hello how are you".  The prompt and output don't matter what matters is the LOGPROB value which is the sum of the log of the completion token sequence probabilities.  If its identical the results were bit exact.

Start with no prompt cache and get the reference LOGPROB:

PROBS=1 CACHE=0 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.4457244873046875

Now turn on cache.  Result is expected to be identical since the prompt has not been cached yet.

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.4457244873046875

As expected, result was identical indicating bit-exact result.  Now submit the same prompt again, this time it is expectred to be different since all token gens are not done with batch size 1:

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.4836981296539307

As expected, result changed.  Now do another run with cache on to see if its consistent, it should be.

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.4836981296539307

As expected, result was consistent.  Now shut off the prompt cache, result should immediately return
to original:

PROBS=1 CACHE=0 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.4457244873046875

As expected, result returned to original no-cache result.

To fix this problem I tested a patch where I force the token generation to be batch size 1 whether prompt cache is on or off.  The unified diff is as follows:

`start diff`
diff -rbEu llama.cpp.old/examples/server/server.cpp llama.cpp/examples/server/server.cpp
--- llama.cpp.old/examples/server/server.cpp    2024-06-03 17:40:16.678938858 -0400
+++ llama.cpp/examples/server/server.cpp        2024-06-04 19:52:34.893251578 -0400
@@ -268,7 +268,9 @@
     std::string *stopping_word;

     // sampling
-    llama_token sampled;
+    llama_token sampled,start_token;
+    size_t start_pos;
+    bool have_start_token=false;
     struct llama_sampling_params sparams;
     llama_sampling_context ** ctx_sampling = nullptr;
     int ctx_sampling_beams;
@@ -1851,6 +1853,7 @@

         slot.command = SLOT_COMMAND_LOAD_PROMPT;
         slot.prompt_tokens.clear();
+	slot.have_start_token=false;

         LOG_INFO("slot is processing task", {
             {"id_slot", slot.id},
@@ -3430,11 +3433,24 @@

                     // entire prompt has been processed - start decoding new tokens
                     if (slot.n_past == slot.n_prompt_tokens) {
-                        slot.state   = SLOT_STATE_PROCESSING;
-                        slot.command = SLOT_COMMAND_NONE;
+                       if (slot.have_start_token) {
+                          llama_batch_add(batch, slot.start_token, slot.start_pos, { slot.id + 1 }, true);
+                          slot.have_start_token=false;
+                          }

                         GGML_ASSERT(batch.n_tokens > 0);

+                       if (batch.n_tokens > 1) {
+                          slot.start_token = batch.token[batch.n_tokens-1];
+                          slot.start_pos = batch.pos[batch.n_tokens-1];
+                          slot.have_start_token=true;
+                          batch.n_tokens--;
+                          }
+                       else {
+                          slot.state   = SLOT_STATE_PROCESSING;
+                          slot.command = SLOT_COMMAND_NONE;
+                          }
+
                         // extract the logits only for the last token
                         batch.logits[batch.n_tokens - 1] = true;

`end diff`

Running with this patch on the same prompt and cache on/off sequencing produces the following:

PROBS=1 CACHE=0 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.5299837589263916

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.5299837589263916

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.5299837589263916

PROBS=1 CACHE=1 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.5299837589263916

PROBS=1 CACHE=0 lm hello how are you?
 Hello! I'm an AI, so I don't have feelings, but I'm here and ready to help you. How can I assist you today?
LOGPROB=-2.5299837589263916

Now the result is identical independent of CACHE on/off, but still different from the first two cases since processing sequence for kv fill is not identical to either CACHE on or off processing in the unpatched server.  However, all token gens are guaranteed to be done with batch size 1 with the patch as long as nslots is configured to be 1 so processing is the same whether prompt cache is on or off with the patch.

**Comment by JohannesGaessler** [comment]
My perspective is this: `cache_prompt` is an option for better performance so by default the implementation details should be done in such a way that the performance is as high as possible. You can make the results bit-for-bit identical for single-turn conversations with the changes you described but then you would only be able to cache the prompt in chunks of the physical batch size. The physical batch size is by default 512 so any prompts shorter than that cannot be cached at all and for prompts longer than that you can estimate that you will still need to reprocess ~256 tokens on average (but with an awkward batch size and a filled context so these will take disproportionally longer).

And this would still not work for multi-turn generations where there are also generated tokens mixed in. You would not be able to cache those either if you want to guarantee bit-for-bit identical results so in those scenarios you would need to reprocess the entire generation from previous turns as well.

So when I said that the issue is "fundamentally unfixable" I was specifically referring to the performance-optimal implementation.

**Comment by steampunque** [comment]
>  You can make the results bit-for-bit identical for single-turn conversations with the changes you described but then you would only be able to cache the prompt in chunks of the physical batch size.

That is not how my patch works.  It processes the whole prompt in chunks of physical batch size except for the last partial chunk where it leaves one token left so it doesn't simultaneously generate on the last partial chunk.

> The physical batch size is by default 512 so any prompts shorter than that cannot be cached at all and for prompts longer than that you can estimate that you will still need to reprocess ~256 tokens on average (but with an awkward batch size and a filled context so these will take disproportionally longer).

No, no tokens are reprocessed.  The processing goes B B B .. P-1 1 1 1 1 1 where B is the batch size and P is the last full partial batch size, P<=B.

>And this would still not work for multi-turn generations 

This is a valid point because Processing 1 1 1 1 1 ..... 1  during token gen is going to give a different result from processing them in B B B... P-1 during next turn prompt processing, but it should also be able to be fixed if desired.

> So when I said that the issue is "fundamentally unfixable" I was specifically referring to the performance-optimal implementation. What I should have said is: "It's fundamentally unfixable unless you were to reimplement it in a way that is suboptimal for performance."

Single prompt is both fixable and performance-optimal, while multi turn cannot be fixed without suboptimal performance, though it could in theory also be fixed by using 1 1 1 ... 1 batch size on the prompt processing/caching of the tokens following the initial prompt.  I will give a quick look into supporting that option also since it may be desired to at least have the option of deterministic results for both single and multiturn prompting.

**Comment by JohannesGaessler** [comment]
>That is not how my patch works. It processes the whole prompt in chunks of physical batch size except for the last partial chunk where it leaves one token left so it doesn't simultaneously generate on the last partial chunk.

Okay, it seems I misunderstood. Yes, something like that should work and the performance impact of generating one token more or less should be negligible.

But consider that for single-turn generations the user can also add or remove arbitrary tokens at the end of the prompt. So I think it will still be necessary to re-process if two prompts are different but share the same prefix.

**Comment by steampunque** [comment]
> > That is not how my patch works. It processes the whole prompt in chunks of physical batch size except for the last partial chunk where it leaves one token left so it doesn't simultaneously generate on the last partial chunk.
> 
> Okay, it seems I misunderstood. Yes, something like that should work and the performance impact of generating one token more or less should be negligible.
> 
> But consider that for single-turn generations the user can also add or remove arbitrary tokens at the end of the prompt. So I think it will still be necessary to re-process if two prompts are different but share the same prefix.

Last partial batch in prompt would need reprocessing, agree.

**Review by ggerganov — APPROVED** [review]

**Comment by JohannesGaessler** [comment]
I'm merging this PR as-is because while it seems that you could in principle fix some of the nondeterminism issues with negligible performance penalties there is (as of yet) no implementation for this. So the documentation change does accurately reflect the behavior on master until and unless there is a change to the code.
